EventLog.md

```diff
@@ -0,0 +1,397 @@
+
+  private def ts(i: Instant): Timestamp =
+    Timestamp.from(i)
+
+  private def selectAllObserveStage(start: Instant, end: Instant): ConnectionIO[List[EventLog.Event]] =
+    sql"""
+      SELECT timestamp,
+             lifecycle,
+             observe_stage,
+             sequence_id
+        FROM log_observe_stage
+       WHERE timestamp BETWEEN ${ts(start)} AND ${ts(end)}
+    ORDER BY timestamp
+    """.query[(Timestamp, Lifecycle, Stage, String)].map { case (t,l,s,i) =>
+      val timestamp = t.toInstant
+      val sid       = Sequence.Id.unsafeFromString(i)
+      val start     = l == Lifecycle.Start
+
+      s match {
+        case Slew   => start ? (EventStartSlew(timestamp, sid):     Event) | EventEndSlew(timestamp, sid)
+        case Visit  => start ? (EventStartVisit(timestamp, sid):    Event) | EventEndVisit(timestamp, sid)
+        case Seq    => start ? (EventStartSequence(timestamp, sid): Event) | EventEndSequence(timestamp, sid)
+      }
+    }.list
+
+  private def selectAllObserveControl(start: Instant, end: Instant): ConnectionIO[List[EventLog.Event]] =
+    sql"""
+      SELECT timestamp,
+             observe_ctrl,
+             sequence_id,
+             why
+        FROM log_observe_ctrl
+       WHERE timestamp BETWEEN ${ts(start)} AND ${ts(end)}
+    ORDER BY timestamp
+    """.query[(Timestamp, Ctrl, String, Option[String])].map { case (t,c,i,w) =>
+      val timestamp = t.toInstant
+      val sid       = Sequence.Id.unsafeFromString(i)
+
+      c match {
+        case Ctrl.Abort    => EventAbortObserve(   timestamp, sid, w): Event
+        case Ctrl.Continue => EventContinueObserve(timestamp, sid, w): Event
+        case Ctrl.Pause    => EventPauseObserve(   timestamp, sid, w): Event
+        case Ctrl.Stop     => EventStopObserve(    timestamp, sid, w): Event
+      }
+    }.list
+
+  private def selectAllObserveIntegration(start: Instant, end: Instant): ConnectionIO[List[EventLog.Event]] =
+    sql"""
+      SELECT timestamp,
+             lifecycle,
+             sequence_id,
+             step
+        FROM log_observe_int
+       WHERE timestamp BETWEEN ${ts(start)} AND ${ts(end)}
+    ORDER BY timestamp
+    """.query[(Timestamp, Lifecycle, String, Int)].map { case (t,l,i,s) =>
+      val timestamp = t.toInstant
+      val sid       = Sequence.Id.unsafeFromString(i)
+
+      l match {
+        case Lifecycle.Start => EventStartIntegration(timestamp, sid, s): Event
+        case Lifecycle.End   => EventEndIntegration(  timestamp, sid, s): Event
+      }
+    }.list
+
+  def selectAll(start: Instant, end: Instant): ConnectionIO[List[EventLog.Event]] =
+    for {
+      c <- selectAllObserveControl(start, end)
+      i <- selectAllObserveIntegration(start, end)
+      s <- selectAllObserveStage(start, end)
+    } yield (c ++ i ++ s).sortBy(_.timestamp)
+
+}
+````
+
+I'm just combining the data collected from the three tables in Scala code.  Maybe there's a more efficient way to do this in a query?
```

Comment:
You can use a `UNION` query which will pull in the step and why as `Option`s and then pattern match using `None` for cases where they can't apply. You'll end up with a final "this doesn't make sense" case that can just throw an exception. But if you're doing this it might be worth smashing into a single table as described above.